### PR TITLE
Add text for low-level vs high-level APIs (related to #117)

### PR DESCRIPTION
**Pull request by LeaVerou**
This is my first, rough draft of text that addresses #117.

Ideally it needs more examples, but I figured I'd push the text while I brainstorm these.

A lot of this prose is based on [minutes](https://github.com/w3ctag/meetings/blob/gh-pages/2020/telcons/05-04-minutes.md#high-level-or-low-level-apis---hober-alice-dbaron-atanassov) from a past breakout, so I've invited all four participants of that conversation to review, including TAG alumni @alice & @dbaron.

***
<a href="https://pr-preview.s3.amazonaws.com/w3ctag/design-principles/pull/291.html" title="Last updated on Jun 21, 2021, 12:11 PM UTC (0f338e9)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3ctag/design-principles/291/6922690...0f338e9.html" title="Last updated on Jun 21, 2021, 12:11 PM UTC (0f338e9)">Diff</a>

### CODE REVIEW

**Comment by LeaVerou** [comment]
I was a little unsure I fully understand this bit well (which is why it's basically worded the same as it was in the minutes):

> Whenever possible, design high level APIs in terms of building blocks over lower level APIs. Note that it may not be possible to expose the lower level building blocks. That may guide decisions on how high level the API needs to be.

So I reached out to @hober for clarification. The gist is that even when the low level blocks can't be exposed, it's still a good idea to define high level APIs in terms of low level ones. The EWM left some people with the impression that any underlying thing must be exposed as API if any platform feature were to depend on it, and we both agree this is incorrect. 
We also agreed that we should include some guidance on when these underlying things shouldn’t be exposed, otherwise it’s entirely subjective. The low hanging fruit there is security & privacy reasons, but are there others?

File: index.bs [hunk 1]
```diff
@@ -269,6 +269,45 @@ See also:
...
+[was recommended by the TAG](https://github.com/w3ctag/design-reviews/issues/399)
+over a Font Enumeration API (low level API)
+as it addresses the bulk of use cases,
+without the fingerprinting concerns that accompany
+a general Font Enumeration API.
+
+</div>
+
+Low-level APIs afford authors room for experimentation
+so that what the high level APIs can organically emerge
```
Comment (hober) [inline index.bs:295]:
```suggestion
so that high level APIs can organically emerge
```

File: index.bs [hunk 2]
```diff
@@ -269,6 +269,45 @@ See also:
...
+from usage patterns over time.
+They also provide an escape hatch when the higher-level API
+is not adequate for the use case at hand.
+
+Whenever possible, design high level APIs in terms of
+building blocks over lower level APIs.
+Note that it may not be possible to expose the lower level building blocks.
```
Comment (hober) [inline index.bs:302]:
```suggestion
Note that it isn't always possible to expose the lower level building blocks as API themselves.
```

File: index.bs [hunk 3]
```diff
@@ -269,6 +269,45 @@ See also:
...
+</div>
+
+Low-level APIs afford authors room for experimentation
+so that what the high level APIs can organically emerge
+from usage patterns over time.
+They also provide an escape hatch when the higher-level API
+is not adequate for the use case at hand.
+
+Whenever possible, design high level APIs in terms of
+building blocks over lower level APIs.
```
Comment (hober) [inline index.bs:301]: I think maybe this sentence should be moved after the next one? So that the reader knows *why* this is the case?

**Review by hober — APPROVED** [review]
Looks good to me. I noted several nits which may or may not be worth adopting.

File: index.bs (hunk 3 above)
Comment (LeaVerou) [reply index.bs:301]: Just pushed a change (somewhat larger than what you were suggesting), thoughts?

**Comment by LeaVerou** [comment]
@hober any thoughts on my changes in response to your feedback? Let's try to get this merged soon 😄

File: index.bs [hunk 4]
```diff
@@ -269,6 +269,49 @@ See also:
...
+However, high level APIs should be designed in terms of
+building blocks over lower level APIs
+whenever possible.
+This may guide decisions on how high level the API needs to be.
+
+A well-layered solution should ensure continuity
+of the ease-of-use vs power tradeoff curve
+and avoid sharp cliffs
+where a small amount of incremental use case complexity
+results in a disproportionately large increase of code complexity.
```
Comment (cynthia) [inline index.bs:313]: I might remove the word "disproportionately" - "large" conveys the message well enough, and the word isn't friendly towards non-native speakers.

File: index.bs [hunk 5]
```diff
@@ -269,6 +269,49 @@ See also:
...
+intervene in various ways [on behalf of the user](#priority-of-constituencies),
+such as to ensure accessibility, privacy, or usability.
+
+<div class=example id=font-enumeration>
+
+A font picker (high level API)
+[was recommended by the TAG](https://github.com/w3ctag/design-reviews/issues/399)
+over a Font Enumeration API (low level API)
+as it addresses the bulk of use cases,
+without the fingerprinting concerns that accompany
```
Comment (atanassov) [inline index.bs:289]: I like the example but your leading point to it was about how good (high level API) are as they provide accessibility, privacy etc. You should add a sentence about Font Picker being accessible, privacy preserving etc.

File: index.bs [hunk 6]
```diff
@@ -269,6 +269,49 @@ See also:
...
+building blocks over lower level APIs
+whenever possible.
+This may guide decisions on how high level the API needs to be.
+
+A well-layered solution should ensure continuity
+of the ease-of-use vs power tradeoff curve
+and avoid sharp cliffs
+where a small amount of incremental use case complexity
+results in a disproportionately large increase of code complexity.
+
```
Comment (atanassov) [inline index.bs:314]: After reading the entire section few times I find myself still unsure about what the guidance is. My recollection from our discussions was that generally we recommend a layered approach to building API and starting by exposing the highest level possible. That allows UAs to take care of all the things you mentioned about a11y, privacy etc., validate scope and functionality and still have a path forward to allow extending by adding lower levels of APIs. It is also much easier to remove few APIs (high level presumably) vs many API.

**Comment by LeaVerou** [comment]
After @atanassov's comments, I wonder if this should actually be merged with https://github.com/w3ctag/design-principles/pull/306 which is basically about the same thing but provides more clear guidance.

File: index.bs [hunk 7]
```diff
@@ -269,6 +269,49 @@ See also:
...
+such as to ensure accessibility, privacy, or usability.
+
+<div class=example id=font-enumeration>
+
+A font picker (high level API)
+[was recommended by the TAG](https://github.com/w3ctag/design-reviews/issues/399)
+over a Font Enumeration API (low level API)
+as it addresses the bulk of use cases,
+without the fingerprinting concerns that accompany
+a general Font Enumeration API.
```
Comment (cynthia) [inline index.bs:290]:
```suggestion
a general Font Enumeration API, while preserving accessibility and privacy.
```

File: index.bs (hunk 5 above)
Comment (LeaVerou) [reply index.bs]: There was already a mention of accessibility and privacy that right under your comment, but I reworded to emphasize. What do you think?

File: index.bs (hunk 6 above)
Comment (torgo) [reply index.bs:317]: @atanassov in discussion this morning we feel the best approach is to land the PR as written and then address the issues you've raised in a subsequent PR.

**Pull request merged**